Allow cacheAfter for inputs that are outputs

jacobhinkle opened the pull request:
Fixes #2372

jacobhinkle committed:
Allow cacheAfter for inputs that are outputs if they are used

naoyam commented on third_party/nvfuser/test/test_gpu3.cpp:
```diff
@@ -6642,6 +6642,70 @@ TEST_F(NVFuserTest, FusionIssue2077_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  auto tx = TensorViewBuilder()
+                .ndims(5)
+                .contiguity(std::vector<bool>(5, true))
+                .dtype(DataType::Float)
+                .build();
```
You could just do `makeContigTensor(5)`.

naoyam commented on third_party/nvfuser/test/test_gpu3.cpp:
```diff
@@ -6642,6 +6642,70 @@ TEST_F(NVFuserTest, FusionIssue2077_CUDA) {
+  auto eager_diff = x - mean.view({1, 1, 1, 1, -1});
+  auto eager_invstd = at::rsqrt(var + eps).view({1, 1, 1, 1, -1});
+
+  ASSERT_TRUE(at::allclose(cg_outputs[0], eager_diff * eager_invstd));
```
Can you use `testValidate`?

jacobhinkle replied:
Good call. I tried converting to use testValidate instead of only checking that the first output matches. This means checking `mean` against `cg_outputs[1]` as well, which is actually failing since it seems the `Fusion` is outputting an empty 0-dimensional second output (`tmean`). Somehow `addOutput(tmean)` is giving us a tensor with `numel()==0` instead of the. I'm investigating this now with a simpler test:
```cpp
TEST_F(NVFuserTest, FusionIssue2372InputOutput_CUDA) {  Fusion fusion;             
  FusionGuard fg(&fusion);                                                     
                                                                            
  auto x = makeContigTensor(1, DataType::Float);
  fusion.addInput(x);
  
  fusion.addOutput(x);
  
  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
                                                   
  at::manual_seed(0);                                                                                                                                                         
  int C = 2;                                                  
  at::Tensor x_at = at::randn({C}, options);              
                                                          
  std::vector<at::IValue> inputs = {x_at};   
                                                              
  auto lparams = schedulePointwise(&fusion, inputs);
  /*
  C++ exception with description "largest_out != nullptr INTERNAL ASSERT FAILED at "../third_party/nvfuser/csrc/scheduler/pointwise.cpp":101, please report a bug to PyTorch.
Exception raised from getPointwiseHeuristics at ../third_party/nvfuser/csrc/scheduler/pointwise.cpp:101 (most recent call first):
  */
                                                                             
  FusionExecutor fe;
  fe.compileFusion(&fusion, inputs, lparams);
  auto cg_outputs = fe.runFusion(inputs, lparams);

  ASSERT_TRUE(at::allclose(cg_outputs[0], x_at));
}
```
This is a different error since the code in the current test does schedule, but it seems related that the scheduler doesn't see any owned outputs (`HeuristicSummaryEntry::data_ptr_` is `nullptr`). I'm currently comparing this to how aliasing works to try and get an idea how to fix this.

naoyam replied:
The schedulers typically use one tensor as a reference of transformations, and it looks like the pointwise scheduler fails to find any reference tensor, likely because of the structure of the fusion.

jacobhinkle replied:
`testValidate` actually skips aliased outputs. So I kept the explicit assert but also added `testValidate` for the other outputs.

naoyam commented on third_party/nvfuser/csrc/tensor_view.cpp:
```diff
@@ -1305,10 +1305,10 @@ TensorView* TensorView::cacheAfter(c10::optional<LoadStoreOpType> cache_op) {
   // Get all the uses for this Tensorview
   TORCH_CHECK(
-      !isFusionOutput(),
+      !(uses().empty() && isFusionOutput()),
```
is `isFusionOutput` necessary? If a tensor is not used, `cacheAfter` doesn't make any sense.

FYI, if a tensor is not used and is not a fusion output, it is usually not computed as it isn't live, except for when a tensor is an output of multi-output expressions. When only part of the output tensors are used, the unused outputs are still computed but their `uses()` should be empty.

jacobhinkle replied:
You're right. I left it there since the check was there originally. But probably the original check was just an approximation for `uses().empty()` anyway. Will remove `isFusionOutput()`.

naoyam reviewed (commented):
Mostly looks good. Just some minor comments.

jacobhinkle committed:
Use makeContigTensor in test FusionIssue2372_CUDA

jacobhinkle committed:
Replace addOutput(input) with call to aliasOutputToInput

This commit alters the order of aliasOutputToInput slightly to give
some state that can be checked inside addOutput in order to avoid
infinite recursion.

jacobhinkle commented on third_party/nvfuser/csrc/fusion.cpp:
```diff
@@ -686,13 +692,14 @@ void Fusion::aliasOutputToInput(Val* output, Val* input) {
     output = castOp(input->getDataType().value(), output);
   }
-  // TODO: output should be marked at the end of fusion definition #1488
-  addOutput(output);
 
   TORCH_INTERNAL_ASSERT(
       isAliasCompatible(input, output),
       "The input and output values are not alias-compatible.");
   io_alias_[output] = input;
+
+  // TODO: output should be marked at the end of fusion definition #1488
+  addOutput(output);
```
@jjsjann123 I switched the order these lines so that `io_alias_[output]` can be checked inside `addOutput` to avoid infinite recursion. I think this is kosher but just letting you know.

jjsjann123 replied:
Appreciate the heads up. This should be fine~~

jacobhinkle committed:
Only check for uses (not isFusionOutput) in cacheAfter

naoyam commented on third_party/nvfuser/csrc/fusion.cpp:
```diff
@@ -226,6 +226,12 @@ void Fusion::addOutput(Val* output) {
+    // aliasOutputToInput
+    aliasOutputToInput(output, output);
+    return;
+  }
```
I'm afraid I have no idea what the comment above means or if it's still relevant. 
@jjsjann123, does this change make sense?

jjsjann123 replied:
hmm... good question!

Did this actually work?! Since we are adding a short-cut for aliased output, looks like we are skipping
`  outputs_.push_back(output);`. Doesn't this remove the kernel arguments?

I'm not sure if we actually care about setting memory type of tv any more, since we don't need to allocate anything.

jacobhinkle replied:
We actually don't skip that part, since `aliasOutputToInput` will call `addOutput` again, and this time find `output` in `io_alias_`, so this if block will not fire.

jjsjann123 replied:
Oops. sorry I mis-read this one. Looking good to me.

naoyam reviewed (approved):
LGTM
